File: app/src/main/java/com/bennyhuo/experimental/coroutines/sample/Ex10_JavaInterop.java
package com.bennyhuo.experimental.coroutines.sample;

import com.bennyhuo.coroutines.sample.CoroutineInterop;
import com.bennyhuo.coroutines.utils.LogKt;

public class Ex10_JavaInterop {
    public static void main(String... args) {
        try {
            LogKt.log(CoroutineInterop.loadString());
            LogKt.log(CoroutineInterop.loadFuture().get());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/bennyhuo/experimental/coroutines/sample/Ex10_JavaInterop.kt
@file:JvmName("CoroutineInterop")
package com.bennyhuo.coroutines.sample

import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.delay
import kotlinx.coroutines.experimental.future.future
import kotlinx.coroutines.experimental.runBlocking

private suspend fun loadAsync(): String {
    delay(10000L)
    return "Hello"
}
private suspend fun loadAsync2(): String {
    delay(1L)
    return "keepon"
}

fun loadString() = runBlocking {
    async{ loadAsync() }.await()
}

//需要 org.jetbrains.kotlinx:kotlinx-coroutines-jdk8:0.22.5，Android 上面目前仍不方便使用
fun loadFuture() = future {
    async{ loadAsync2() }.await()
}